tests/personal_data/transactions/test_transactions_manager.py

```diff
@@ -0,0 +1,203 @@
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+import decimal
+import uuid
+
+import pytest
+
+import octobot_trading.enums as enums
+import octobot_trading.errors as errors
+import octobot_trading.constants as constants
+import octobot_trading.personal_data.transactions.types as transaction_types
+
+from tests.exchanges import backtesting_trader, backtesting_config, backtesting_exchange_manager, fake_backtesting
+from tests import event_loop
+
+# All test coroutines will be treated as marked.
+pytestmark = pytest.mark.asyncio
+
+TRANSACTION_SYMBOL = "BTC/USDT"
+TRANSACTION_CURRENCY = "BTC"
+
+
+async def test_initialized(backtesting_trader):
+    _, exchange_manager, _ = backtesting_trader
+    assert exchange_manager.exchange_personal_data.transactions_manager.is_initialized
+
+
+async def test_upsert_transaction_instance(backtesting_trader):
+    _, exchange_manager, _ = backtesting_trader
+    assert len(exchange_manager.exchange_personal_data.transactions_manager.transactions) == 0
+    t_id = str(uuid.uuid4())
+    t_id_2 = str(uuid.uuid4())
+    transaction = transaction_types.RealisedPnlTransaction(
+        exchange_name=exchange_manager.exchange_name,
+        creation_time=exchange_manager.exchange.get_exchange_current_time(),
+        transaction_type=enums.TransactionType.REALISED_PNL,
+        currency=TRANSACTION_CURRENCY,
+        symbol=TRANSACTION_SYMBOL,
+        realised_pnl=constants.ZERO)
+    transaction.set_transaction_id(t_id)
+    transaction_2 = transaction_types.BlockchainTransaction(
+        exchange_name=exchange_manager.exchange_name,
+        creation_time=exchange_manager.exchange.get_exchange_current_time(),
+        currency=TRANSACTION_CURRENCY,
+        transaction_type=enums.TransactionType.BLOCKCHAIN_DEPOSIT,
+        blockchain_type="Test",
+        blockchain_transaction_id=t_id,
+        blockchain_transaction_status=enums.BlockchainTransactionStatus.CONFIRMING,
+        source_address="0x123456789",
+        destination_address=None,
+        quantity=decimal.Decimal(50),
+        transaction_fee=decimal.Decimal(0.1)
+    )
+    transaction_2.set_transaction_id(t_id)
+
+    # succeed to upsert unknown transaction
+    assert len(exchange_manager.exchange_personal_data.transactions_manager.transactions) == 0
+    exchange_manager.exchange_personal_data.transactions_manager.upsert_transaction_instance(transaction)
+    assert len(exchange_manager.exchange_personal_data.transactions_manager.transactions) == 1
+
+    # can't add transaction with the same transaction id without replacing it --> DuplicateTransactionIdError
+    with pytest.raises(errors.DuplicateTransactionIdError):
+        exchange_manager.exchange_personal_data.transactions_manager.upsert_transaction_instance(transaction_2)
+    assert len(exchange_manager.exchange_personal_data.transactions_manager.transactions) == 1
```

Comment:
Here it would be nice to check that the transaction has not been replaced:
`assert list(exchange_manager.exchange_personal_data.transactions_manager.transactions.values())[0] is transaction_1`